server/chat-plugins/helptickets-auto.ts

```diff
@@ -0,0 +1,670 @@
+	const res = await Rooms.Modlog.search('global', search);
+	return res?.results || [];
+}
+
+function closeTicket(ticket: TicketState, msg?: string) {
+	if (!ticket.open) return;
+	ticket.open = false;
+	ticket.active = false;
+	ticket.resolved = {
+		time: Date.now(),
+		by: 'the Artemis AI', // we want it to be clear to end users that it was not a human
+		seen: false,
+		staffReason: '',
+		result: msg || "",
+		note: `Want to learn more about the AI? <a href="https://570628/#post-9056769">Visit the information thread</a>.`,
+	};
+	writeTickets();
+	notifyStaff();
+	const tarUser = Users.get(ticket.userid);
+	if (tarUser) {
+		HelpTicket.notifyResolved(tarUser, ticket, ticket.userid);
+	}
+	// TODO: Support closing as invalid
+	writeStats(`${ticket.type}\t${Date.now() - ticket.created}\t0\t0\tresolved\tvalid\tartemis`);
+}
+
+async function lock(
+	user: User | ID,
+	result: CheckerResult,
+	ticket: TicketState,
+	isWeek?: boolean,
+	isName?: boolean
+) {
+	const id = toID(user);
+	let desc, type;
+	const expireTime = isWeek ? Date.now() + 7 * 24 * 60 * 60 * 1000 : null;
+	if (isName) {
+		if (typeof user === 'object') user.resetName();
+		desc = 'locked your username and prevented you from changing names';
+		type = `locked from talking${isWeek ? ` for a week` : ""}`;
+		await Punishments.namelock(id, expireTime, null, false, result.reason || "Automatically locked due to a user report");
+	} else {
+		type = isWeek ? 'weeknamelocked' : 'namelocked';
+		desc = 'locked you from talking in chats, battles, and PMing regular users';
+		await Punishments.lock(id, expireTime, null, false, result.reason || "Automatically locked due to a user report");
+	}
+	if (typeof user !== 'string') {
+		let message = `|popup||html|${user.name} has ${desc} for ${isWeek ? '7' : '2'} days.`;
+		if (result.reason) message += `\n\nReason: ${result.reason}`;
+		let appeal = '';
+		if (Chat.pages.help) {
+			appeal += `<a href="view-help-request--appeal"><button class="button"><strong>Appeal your punishment</strong></button></a>`;
+		} else if (Config.appealurl) {
+			appeal += `appeal: <a href="${Config.appealurl}">${Config.appealurl}</a>`;
+		}
+		if (appeal) message += `\n\nIf you feel that your lock was unjustified, you can ${appeal}.`;
+		message += `\n\nYour lock will expire in a few days.`;
+		user.send(message);
+	}
+	addModAction(`${id} was ${type} by Artemis. (${result.reason || `report from ${ticket.creator}`})`);
+	globalModlog(
+		`${isWeek ? 'WEEK' : ""}${isName ? "NAME" : ""}LOCK`, id,
+		(result.reason || `report from ${ticket.creator}`) + (result.proof ? ` PROOF: ${result.proof}` : "")
+	);
+}
+
+export const actionHandlers: {
+	[k: string]: (user: User | ID, result: CheckerResult, ticket: TicketState) => string | void | Promise<string | void>,
+} = {
+	forcerename(user, result, ticket) {
+		if (typeof user === 'string') return; // they can only submit users with existing userobjects anyway
```

Comment:
Don't we have offline forcerenames now?

Reply:
We do, but helptickets only allow for reporting of online users as a guard against typos.

Reply:
In that case you could look up the user object and do an online forcerename; why reject IDs?

Reply:
The inapname checker already does that lookup. IDs should never be here, so this is just to appease TS.